# Guided Project 1 - Apps - What ones do we want to make?

This project piece is the culmination of of the first section of the Dataquest guided learning.

In this scenarion, I am undertaking the role of an analyst working for a company which makes mobile apps for Andriod and IOS Apple devices, to be available on Google Play and the Apple App Store.

The apps which are developed are free and revenue comes from in game adds, so the purpose of this analytical piece is to determine which apps are most profitable in the above sense by determining what type of apps attract the most users and therefore the most add revenue.



Two datasets used:  
- Google Play Store Data: https://dq-content.s3.amazonaws.com/350/googleplaystore.csv
- Apple Play Store Data: https://dq-content.s3.amazonaws.com/350/AppleStore.csv

 

## Step 1 - Opening the datasets

Both datasets need to be opened, read and transformed into a list of lists.

This will be done by:  
- importing the `reader` command from the `CSV` module
- opening the files which are stored locally to on my computer
- stipulating the encoding as `utf8` to aleviate any issues
- reading the files
- converting the read files into three list variables, one containing the entire dataset, one containing just the headers and one with just the data

In [1]:
from csv import reader

# opening the android data #

opened = open(r"C:\Users\andre\OneDrive\Desktop\Python Files\Guided Project - Apps\googleplaystore.csv",encoding="utf8")
read = reader(opened)
android_all_data = list(read)
android_headers = android_all_data[0]
android_rows = android_all_data[1:]

# opening the apple data #

opened = open(r"C:\Users\andre\OneDrive\Desktop\Python Files\Guided Project - Apps\AppleStore.csv",encoding="utf8")
read = reader(opened)
apple_all_data = list(read)
apple_headers = apple_all_data[0]
apple_rows = apple_all_data[1:]




## Step 2 - initial explore of the data

To make exploring the data easier in the first instance, we created a function entitled `explore_data()` which takes in 4 arguements:  
- `dataset` which is intended to be taken as a list of lists
- `start` and `end` which will take the index numbers of the rows we wish to look at
- `rows_columns` which will take a boolean value and be defaulted to `False`

This function will enable us to easily view a select few rows of the input datasets.

In [2]:
def explore_data(dataset, start, end, rows_columns = False):
    for row in dataset[start:end]:
        print(row)
        print('\n')
        
    if rows_columns:
        print('Number of rows: ', len(dataset))
        print('Number of columns: ', len(dataset[0]))
        
print(explore_data(android_all_data, 0, 3, True))

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


Number of rows:  10842
Number of columns:  13
None


In the above we have run the `explore_data` function for first 3 rows for the `android_all_data` data set which does contain the headers/column titles; so the output is the column titles and the first 3 rows of data.

We have also stipulated ` rows_columns` to be `True` so the output also included a count of the rows and columns in the android_all_data dataset. We can see that in total, the dataset contains data for 10,841 apps (note the first row of this dataset is the headers) and 13 columns.

At a quick glance, the columns which could be of use at a later date are:  
- `App`
- `Category`
- `Rating`
- `Reviews`
- `Installs`
- `Type`
- `Price`
- `Genres`

As we have already defined the function `explore_data()` in the section above we do not need to repeat our actions to initially explore the Apple App Store dataset, we simply need to call the fucntion and change the `dataset` parameter to one of the 3 apple lists we created in Step 1.  
  
For the below I will again use the list containing the entirety of the dataset including headers.

In [3]:
print(explore_data(apple_all_data, 0, 3, True))

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


Number of rows:  7198
Number of columns:  16
None


We can see that the Apple App dataset contains information on 7,197 apps (again `row[0]` contains the headers) and holds 16 columns, 3 more than the Google Play dataset.  
  
  The columns which could be of use appear to be:  
  - `track_name`
  -` price`
  - `rating_count_tot`
  - `rating_count_ver`
  - `user_rating`
  - `prime_genre`

## Step 3 - Deleting Incomplete Data - Google Play Store  
  
In this article (https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015) about the google play dataset, it is mentioned that there is a missing `rating` for entry `[10,472]` and a column shift has occured as a result.

Whilst the end result of this section will be to simply remove said row from our dataset using the command `del android_all_data[10472]`, for the purposes of this exercise I am going to pretend that we do not know where this error has occured and need to undertake steps to check all of our rows have the correct amount of data points (13).  
  
I will do this by comparing the number of columns held in the `android_headers` list to the number of entries each row has in the `android_all_data` list of lists.



In [4]:
for row in android_all_data[1:]:       
    if len(row) != len(android_headers):
     
        print('index number: ', android_all_data.index(row))
        print('\n')
        print('row length: ', len(row))
        print('\n')
        print(row)
       
        


index number:  10473


row length:  12


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


As a point of note, the index number returned from the above is 1 greater than the index number being discussed in the link above. Reading the discussion further it seems that the parties involved in the discussion were working on a version of the dataset with the headers removed whereas I was working on the full version which accounts for the discrepancy.


For the sake of thoroughness and again assuming we haven't been given the answer via a public forum, we should run the above check for the Apple dataset as well.  
  
  I will again be running the check on the full dataset.

In [5]:
for row in apple_all_data[1:]:       
    if len(row) != len(apple_headers):
     
        print('index number: ', apple_all_data.index(row))
        print('\n')
        print('row length: ', len(row))
        print('\n')
        print(row)

Running the above code returns no values, therefore confirming that from the perspecitve of this checking method the Apple dataset is 100% accurate.

We now need to remove the corrupt data point from the Google Play data set and will be done by running the code below.

Note, I have turned the below code into a note after I ran it to prevent me from inadvertently running it in the future and removing other complete rows.

In [6]:
#del android_all_data[10473]

In [7]:
print(len(android_all_data))

10841


All being well, if we rerun the checks for the google dataset we should be met with a nil return thereby indicating that the corrupt row has been removed and the rest of the dataset is as we need it.

In [8]:
for row in android_all_data[1:]:       
    if len(row) != len(android_headers):
     
        print('index number: ', android_all_data.index(row))
        print('\n')
        print('row length: ', len(row))
        print('\n')
        print(row)

As hoped, the above has returned nothing so we are in a good position to proceed.

## Step 4 - Removing Duplicates from the data  
  
As with any new dataset part of the initial cleansing process should be to identify and subsiquently remove any and all duplicate entries; so we shall be doing this for both of our datasets by using the Name of the apps in the datasets as our unique identifiers.  
  
  ### Identifying the duplicates  
  To identify the duplicates from the Google Play Store dataset, we will use a series of lists entitled `unique` and `duplicate` and itterate over the full dataset to idenfity what, if any, apps names appear more than once.

In [9]:
duplicate_apps = []
unique_apps = []

for row in android_all_data[1:]:
    if row[0] in unique_apps:    #we know from previous steps that the first column (index 0) of the goolge data contains the apps name
        duplicate_apps.append(row[0])
    else:
        unique_apps.append(row[0])
        
print('Number of duplicates: ', len(duplicate_apps))
print('\n')
print('Duplicate examples:')
for i in duplicate_apps[:15]:
    print(i)        

Number of duplicates:  1181


Duplicate examples:
Quick PDF Scanner + OCR FREE
Box
Google My Business
ZOOM Cloud Meetings
join.me - Simple Meetings
Box
Zenefits
Google Ads
Google My Business
Slack
FreshBooks Classic
Insightly CRM
QuickBooks Accounting: Invoicing & Expenses
HipChat - Chat Built for Teams
Xero Accounting Software


### Removing the duplicates

We can now see what within the full google play store dataset 1181 entries are duplicates and action needs to be taken to remove them. However before we remove them we need to assertain what the deletion criteria is to be to avoid erroneously deleting datapoints.  
  
Lets start by looking at all of the entries for one of the duplicated apps, namely `Instagram`.

In [10]:
print(android_headers)
for row in android_all_data[1:]:
    if row[0] == 'Instagram':
        print(row)
        print('\n')

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']




With the exception of the number of reviews reveived (column 4/ index 3) all the information for the duplicate apps appears to be the same so we will be focusing on using the number of reviews each app has received to remove our duplicates.  
  
Logic dictates that the datapoint with the highest number of reviews is the one we want to keep and work with so the code below will work on that premis.

In [11]:
max_reviews = {}

for row in android_all_data[1:]:
    if (row[0] in max_reviews) and (float(row[3]) < max_reviews[row[0]]):
        max_reviews[row[0]] = float(row[3])
        
    elif row[0] not in max_reviews:
        max_reviews[row[0]] = float(row[3])
        
print('Expected Length: ', len(android_all_data) - 1 - len(duplicate_apps))
print('Actual Length: ', len(max_reviews))

Expected Length:  9659
Actual Length:  9659


The values printed above confrims that the expected length of our soon to be de-duplicated dataset matches the actual length of the list of apps once the duplicates have been removed.  
  
We will now use our `max_reviews` dictionary to assist in removing the duplicated rows from our primary dataset.

In [12]:
android_cleaned = []
android_added = []

for row in android_all_data[1:]:
    if float(row[3]) == max_reviews[row[0]] and row[0] not in android_added:
        android_cleaned.append(row)
        android_added.append(row[0])

In [13]:
explore_data(android_cleaned, 0, 5, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


Number of rows:  9659
Number of columns:  13


The number of rows in our new cleaned dataset matches that of our `max_reviews` dictionary and our expected length, happy days!!

A quick check of the Apple Store data will show that there are no duplicates so the above steps do not need to be repeated.


In [14]:
unique_apps = []
duplicate_apps = []

for row in apple_all_data[1:]:
    if row[0] in unique_apps:
        duplicate_apps.append(row[0])
    else:
        unique_apps.append(row[0])
        
print('Number of duplicates: ', len(duplicate_apps))
print('\n')
print('Duplicate examples:')
for i in duplicate_apps[:15]:
    print(i)

Number of duplicates:  0


Duplicate examples:


## Step 5 - Removing Non English apps
  
Given the universality of both app stores and the fact that we are undertaking this whole project for a company which works in English and only makes apps for English speaking customers, we need to again explore our data to idetify and subsiquently remove any apps which appear to not be for an English speaking audience as we do not want to include them in our analysis to follow.  
  
Using the American Standard Code for Information Interchange (ASCII) as the base of our work, it is known that the charachters commonly used in written English possess an `ord()` number between 0 and 127 so we will use this fact to build a function to check the names of our apps to identify if any non English characters are used.

First we need to build a function that checks a single sting input:

In [15]:
def is_english(string):
    for c in string:
        if ord(c) > 127:
            return False
    return True
    
print(is_english('视剧视剧'))
print(is_english('hello'))
      

False
True


Whilst the above function works as desired, certain characters, such as smilies, have an `ord()` value which falls outside the conventional 0 to 127 range so amendments need to be made to not automatically rule out an english app title with a small amount of characters outside the range.  
  
  For the purposes of this exercise I have decided that an app title can have upto a maximum of 3 non standard characters before it is excluded.

In [16]:
def is_english(string):
    ascii_count = 0
    
    for c in string:
        if ord(c) > 127:
            ascii_count += 1
            
    if ascii_count >3:
        return False
    else:
        return True
    
print(is_english('视剧视剧'))
print(is_english('hello 😜'))


False
True


Now we have a function built with an acceptable tollerance limit for non english characters, we can use this to filter out the non english apps from both datasets, creating two new datasets as we go.

When looking through the Google Play we need to ensure we are using the newly created `android_cleaned` list from the previous section.


In [17]:
android_eng = []
android_non_eng_rem = 0
apple_eng = []
apple_non_eng_rem = 0

for row in android_cleaned:
    if is_english(str(row[0])):
        android_eng.append(row)
    elif is_english(str(row[0])) is False:
        android_non_eng_rem += 1
    
for row in apple_rows:
    if is_english(str(row[1])):
        apple_eng.append(row)
    elif is_english(str(row[0])) is False:
        apple_non_eng_rem += 1
        
print(explore_data(android_eng,0,2, True))
print('Non-english Android Apps remove: ', android_non_eng_rem)
print('\n')
print(explore_data(apple_eng,0,2,True))
print('Non-english Apple Apps remove: ', apple_non_eng_rem)

    

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


Number of rows:  9614
Number of columns:  13
None
Non-english Android Apps remove:  45


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


Number of rows:  6183
Number of columns:  16
None
Non-english Apple Apps remove:  0


## Step 6 - Isolating free apps  
The final cleansing exercise we will undertake for this project is to isolate only the free apps from our previously cleansed dataset, as we are working for a developer who only builds free to instal apps.

In [18]:
android_final = []
apple_final = []

for row in android_eng:
    if row[7] == '0':
        android_final.append(row)
        
for row in apple_eng:
    if row[4] == '0.0':
        apple_final.append(row)
        
  
        
print('android_final length: ', len(android_final))
print('apple_final length: ', len(apple_final))

android_final length:  8862
apple_final length:  3222


## Step 7 - Finding the most common apps by genre

For the sake of ease and so save us continuously having to refer to the early sections of this project, I will run the explore function to return the headers and 2 rows of data for both of our datasets.

In [19]:
print(android_headers)
print('\n')
print(explore_data(android_final, 0, 2))

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


None


In [20]:
print(apple_headers)
print('\n')
print(explore_data(apple_final, 0, 2))

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


None


Firstly, we need to build a function to populate a frequency table and to then sort said table in decending order. 
We will build the frequency tables to show the characteristics of the given column as a percentage instead of an absolute value.

In [22]:
def freq_table(dataset, index):
    new_table = {}
    total = 0
    
    for row in dataset:
        total += 1
        if row[index] in new_table:
            new_table[row[index]] += 1
        else:
            new_table[row[index]] = 1
            
    table_perc = {}
    
    for key in new_table:
        percentage = round(((new_table[key] / total) * 100), 2)
        table_perc[key] = percentage
        
    table_to_display = []
    
    for key in table_perc:
        table_to_display.append((table_perc[key],key))
        
    table_sorted = sorted(table_to_display, reverse = True)
    
    for e in table_sorted:
        print(e[1], ': ', e[0], '%')

print(freq_table(apple_final, -5))


        
            

Games :  58.16 %
Entertainment :  7.88 %
Photo & Video :  4.97 %
Education :  3.66 %
Social Networking :  3.29 %
Shopping :  2.61 %
Utilities :  2.51 %
Sports :  2.14 %
Music :  2.05 %
Health & Fitness :  2.02 %
Productivity :  1.74 %
Lifestyle :  1.58 %
News :  1.33 %
Travel :  1.24 %
Finance :  1.12 %
Weather :  0.87 %
Food & Drink :  0.81 %
Reference :  0.56 %
Business :  0.53 %
Book :  0.43 %
Navigation :  0.19 %
Medical :  0.19 %
Catalogs :  0.12 %
None


In [30]:
def freq_table2(dataset, index):
    new_table = {}
    total = 0
    
    for row in dataset:
        total += 1
        if row[index] in new_table:
            new_table[row[index]] += 1
        else:
            new_table[row[index]] = 1
            
    table_perc = {}
    
    for key in new_table:
        percentage = round(((new_table[key] / total) * 100), 2)
        table_perc[key] = percentage
        
    return table_perc

From the above, we can see that for the Apple Store dataset over half the apps are Games (58.16%) with the next most popular genre being that of Entertainment (7.88%).  

For the Google Play Store dataset there are two columns which we want to explore:  
- `Catagories`
- `Genres`

Lets look at `catagory` first.

In [23]:
print(freq_table(android_final,1))

FAMILY :  18.79 %
GAME :  9.64 %
TOOLS :  8.44 %
BUSINESS :  4.58 %
LIFESTYLE :  3.9 %
PRODUCTIVITY :  3.89 %
FINANCE :  3.7 %
MEDICAL :  3.53 %
SPORTS :  3.42 %
PERSONALIZATION :  3.32 %
COMMUNICATION :  3.25 %
HEALTH_AND_FITNESS :  3.07 %
PHOTOGRAPHY :  2.95 %
NEWS_AND_MAGAZINES :  2.8 %
SOCIAL :  2.66 %
TRAVEL_AND_LOCAL :  2.34 %
SHOPPING :  2.25 %
BOOKS_AND_REFERENCE :  2.14 %
DATING :  1.86 %
VIDEO_PLAYERS :  1.78 %
MAPS_AND_NAVIGATION :  1.4 %
EDUCATION :  1.25 %
FOOD_AND_DRINK :  1.24 %
ENTERTAINMENT :  1.04 %
LIBRARIES_AND_DEMO :  0.94 %
AUTO_AND_VEHICLES :  0.93 %
HOUSE_AND_HOME :  0.84 %
WEATHER :  0.8 %
EVENTS :  0.71 %
ART_AND_DESIGN :  0.68 %
PARENTING :  0.65 %
COMICS :  0.62 %
BEAUTY :  0.6 %
None


Conversly, the spread of the catagories in the google store is much closer than that of the Apple Store. The most common catagory here is that of `Family` which makes up 18.79% of the dataset with `Game` coming in second at 9.64%.  
  
And now to look at the `Genre` column:

In [25]:
print(freq_table(android_final,-4))

Tools :  8.43 %
Entertainment :  6.07 %
Education :  5.35 %
Business :  4.58 %
Productivity :  3.89 %
Lifestyle :  3.89 %
Finance :  3.7 %
Medical :  3.53 %
Sports :  3.46 %
Personalization :  3.32 %
Communication :  3.25 %
Action :  3.1 %
Health & Fitness :  3.07 %
Photography :  2.95 %
News & Magazines :  2.8 %
Social :  2.66 %
Travel & Local :  2.32 %
Shopping :  2.25 %
Books & Reference :  2.14 %
Simulation :  2.04 %
Dating :  1.86 %
Arcade :  1.86 %
Video Players & Editors :  1.78 %
Casual :  1.75 %
Maps & Navigation :  1.4 %
Food & Drink :  1.24 %
Puzzle :  1.13 %
Racing :  0.99 %
Role Playing :  0.94 %
Libraries & Demo :  0.94 %
Auto & Vehicles :  0.93 %
Strategy :  0.91 %
House & Home :  0.84 %
Weather :  0.8 %
Events :  0.71 %
Adventure :  0.68 %
Comics :  0.61 %
Beauty :  0.6 %
Art & Design :  0.6 %
Parenting :  0.5 %
Card :  0.44 %
Casino :  0.43 %
Trivia :  0.42 %
Educational;Education :  0.39 %
Educational :  0.37 %
Board :  0.37 %
Education;Education :  0.34 %
Word :  0.2

Whilst the difference between the `catagory` and `genres` fields aren't immediately clear, it is apparant that `genres` is much more granular than `catagory` as it has significantly more unique values.  
  
Given this exercise is a high level one, we shall use `catagory` going forward.


## Step 8 - Finding the average number of installs by genre

Just because the Apple App Store is dominated by `Game`, this doesn't necisarrily mean that this is the most popular catagory for the users. To explore this premice further we will try and calculate the average number of installs per catagory.  
  
  Whilst the Apple App Store data set does not contain a datapoint for th enumber of installs, we can use the number of ratings received as an alternative.

In [31]:
genres_apple = freq_table2(apple_final,-5)

for genre in genres_apple:
    total = 0
    len_genre = 0
    
    for row in apple_final:
        genre_row = row[-5]
        if genre_row == genre:
            total += float(row[5])
            len_genre += 1
            
    average_number_ratings = round(total / len_genre, 2)
    
    print(genre, ":", average_number_ratings)   
     

Social Networking : 71548.35
Photo & Video : 28441.54
Games : 22788.67
Music : 57326.53
Reference : 74942.11
Health & Fitness : 23298.02
Weather : 52279.89
Utilities : 18684.46
Travel : 28243.8
Shopping : 26919.69
News : 21248.02
Navigation : 86090.33
Lifestyle : 16485.76
Entertainment : 14029.83
Food & Drink : 33333.92
Sports : 23008.9
Book : 39758.5
Finance : 31467.94
Education : 7003.98
Productivity : 21028.41
Business : 7491.12
Catalogs : 4004.0
Medical : 612.0


At face value `Navigation` apps hold the highest number of average reviews. However, when using a mean average calculation as we did above they are susceptible to being skewed.  
  
As evidenced below, in the Apple Store dataset there are a total of 6 naviagtion apps and 4 of those have less than 15,000 reviews.

In [33]:
for app in apple_final:
    if app[-5] == "Navigation":
        print(app[1], ":", app[5])

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


**Full disclosure, I am losing steam a tad with the full write up part of this project, mainly as it is taking up too much time.**
  
**From here on in I will just be doing the coding part of the guided project as I am keen to move on to the intermediate intro to Python course.**


In [34]:
freq_table(android_final,5)

1,000,000+ :  15.75 %
100,000+ :  11.55 %
10,000,000+ :  10.51 %
10,000+ :  10.22 %
1,000+ :  8.41 %
100+ :  6.92 %
5,000,000+ :  6.82 %
500,000+ :  5.57 %
50,000+ :  4.77 %
5,000+ :  4.51 %
10+ :  3.54 %
500+ :  3.25 %
50,000,000+ :  2.28 %
100,000,000+ :  2.12 %
50+ :  1.92 %
5+ :  0.79 %
1+ :  0.51 %
500,000,000+ :  0.27 %
1,000,000,000+ :  0.23 %
0+ :  0.05 %
0 :  0.01 %


The ratings have been rounded, catagorised and include a + and , symbol which will need to be removed to enable us to calculate the average.

In [36]:
catagories_android = freq_table2(android_final,1)

for cat in catagories_android:
    total = 0
    len_catagory = 0
    
    for app in android_final:
        catagory_app = app[1]
        if catagory_app == cat:
            number_installs = app[5]
            number_installs = number_installs.replace('+', '')
            number_installs = number_installs.replace(',', '')
            total += float(number_installs)
            len_catagory += 1

    average_installs = round(total / len_catagory, 2)
    print(cat, ':', average_installs)
        

ART_AND_DESIGN : 1905351.67
AUTO_AND_VEHICLES : 647317.82
BEAUTY : 513151.89
BOOKS_AND_REFERENCE : 8767811.89
BUSINESS : 1704192.34
COMICS : 817657.27
COMMUNICATION : 38326063.2
DATING : 854028.83
EDUCATION : 3057207.21
ENTERTAINMENT : 19428913.04
EVENTS : 253542.22
FINANCE : 1387692.48
FOOD_AND_DRINK : 1924897.74
HEALTH_AND_FITNESS : 4167457.36
HOUSE_AND_HOME : 1313681.91
LIBRARIES_AND_DEMO : 638503.73
LIFESTYLE : 1437816.27
GAME : 13006872.89
FAMILY : 4371709.12
MEDICAL : 107167.23
SOCIAL : 23253652.13
SHOPPING : 7036877.31
PHOTOGRAPHY : 17805627.64
SPORTS : 4274688.72
TRAVEL_AND_LOCAL : 13984077.71
TOOLS : 10695245.29
PERSONALIZATION : 5201482.61
PRODUCTIVITY : 16772838.59
PARENTING : 542603.62
WEATHER : 5074486.2
VIDEO_PLAYERS : 24790074.18
NEWS_AND_MAGAZINES : 9549178.47
MAPS_AND_NAVIGATION : 4056941.77
